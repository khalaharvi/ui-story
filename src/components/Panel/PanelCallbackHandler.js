import { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * Helper wrapper for <xxxxFacet /> components wrapped in with <Panel /> component.
 * It is reponsible to update 'canRefine' state in <xxxxFacet />'s parent <Panel /> component.
 *
 * Usage:
 * - Export optionally <Panel />-wrapped <xxxxFacet /> component like so:
 *    <PanelCallbackHandler>
 *      <xxxxFacet />
 *    </PanelCallbackHandler>
 *
 * - Wrap <PanelCallbackHandler>-wrapped <xxxxFacet /> component with <Panel />
 *    <Panel>
 *      <xxxxFacet />
 *    </Panel>
 *
 * From: https://github.com/algolia/react-instantsearch/blob/v5.0.1/packages/react-instantsearch/src/components/PanelCallbackHandler.js
 */
class PanelCallbackHandler extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    canRefine: PropTypes.bool.isRequired
  };

  static contextTypes = {
    setCanRefine: PropTypes.func
  };

  componentWillMount() {
    if (this.context.setCanRefine) {
      this.context.setCanRefine(this.props.canRefine);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.context.setCanRefine && this.props.canRefine !== nextProps.canRefine) {
      this.context.setCanRefine(nextProps.canRefine);
    }
  }

  render() {
    return this.props.children;
  }
}

export default PanelCallbackHandler;
