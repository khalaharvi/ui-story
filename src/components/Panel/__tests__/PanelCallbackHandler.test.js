import React from 'react';
import { shallow } from 'enzyme';

import PanelCallbackHandler from '../PanelCallbackHandler';

// From: https://github.com/algolia/react-instantsearch/blob/v5.0.1/packages/react-instantsearch/src/components/PanelCallbackHandler.test.js
describe('<PanelCallbackHandler />', () => {
  it('renders correctly', () => {
    const props = {
      canRefine: true
    };

    const wrapper = shallow(
      <PanelCallbackHandler {...props}>
        <div>Hello content</div>
      </PanelCallbackHandler>
    );

    expect(wrapper).toMatchSnapshot();
  });

  describe('willMount', () => {
    it('should call setCanRefine when the context is given', () => {
      const props = {
        canRefine: true
      };

      const context = {
        setCanRefine: jest.fn()
      };

      shallow(
        <PanelCallbackHandler {...props}>
          <div>Hello content</div>
        </PanelCallbackHandler>,
        { context }
      );

      expect(context.setCanRefine).toHaveBeenCalledTimes(1);
      expect(context.setCanRefine).toHaveBeenCalledWith(true);
    });

    it('should not throw when the context is not given', () => {
      const props = {
        canRefine: true
      };

      const trigger = () =>
        shallow(
          <PanelCallbackHandler {...props}>
            <div>Hello content</div>
          </PanelCallbackHandler>
        );

      expect(() => trigger()).not.toThrow();
    });
  });

  describe('willReceiveProps', () => {
    it('should call setCanRefine when the context is given', () => {
      const props = {
        canRefine: true
      };

      const context = {
        setCanRefine: jest.fn()
      };

      const wrapper = shallow(
        <PanelCallbackHandler {...props}>
          <div>Hello content</div>
        </PanelCallbackHandler>,
        { context }
      );

      wrapper.setProps({ canRefine: false });

      expect(context.setCanRefine).toHaveBeenCalledTimes(2);
      expect(context.setCanRefine).toHaveBeenLastCalledWith(false);
    });

    it('should not call setCanRefine when the nextProps is the same', () => {
      const props = {
        canRefine: true
      };

      const context = {
        setCanRefine: jest.fn()
      };

      const wrapper = shallow(
        <PanelCallbackHandler {...props}>
          <div>Hello content</div>
        </PanelCallbackHandler>,
        { context }
      );

      wrapper.setProps({ canRefine: true });

      expect(context.setCanRefine).toHaveBeenCalledTimes(1);
    });

    it('should not throw when the context is not given', () => {
      const props = {
        canRefine: true
      };

      const trigger = () => {
        const wrapper = shallow(
          <PanelCallbackHandler {...props}>
            <div>Hello content</div>
          </PanelCallbackHandler>
        );

        wrapper.setProps({ canRefine: false });
      };

      expect(() => trigger()).not.toThrow();
    });
  });
});
