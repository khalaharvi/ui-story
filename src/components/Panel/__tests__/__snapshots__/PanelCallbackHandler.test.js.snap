// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PanelCallbackHandler /> renders correctly 1`] = `
<div>
  Hello content
</div>
`;
